### PR TITLE
Get map of interface name to index

### PR DESCRIPTION
**Pull request by jroggeman**
Needs more thorough testing

### CODE REVIEW

File: netsh/netsh.go [hunk 1]
```diff
@@ -32,6 +33,8 @@ type Interface interface {
 	GetInterfaceByIP(ipAddr string) (Ipv4Interface, error)
 	// Enable forwarding on the interface (name or index)
 	EnableForwarding(iface string) error
+	// Gets a map of interface names to their indexes
+	GetInterfaceNameToIndexMap() (map[string]int, error)
```
Comment (rakelkar) [inline netsh/netsh.go]: >GetInterfaceNameToIndexMap [](start = 1, length = 26)

not useful as public

File: netsh/netsh.go [hunk 2]
```diff
@@ -85,6 +89,10 @@ func (runner *runner) GetInterfaces() ([]Ipv4Interface, error) {
 	var currentInterface Ipv4Interface
 	quotedPattern := regexp.MustCompile("\\\"(.*?)\\\"")
 	cidrPattern := regexp.MustCompile("\\/(.*?)\\ ")
+
+	// Retrieve interface index map
+	indexMap, err := runner.GetInterfaceNameToIndexMap()
```
Comment (rakelkar) [inline netsh/netsh.go]: >err [](start = 11, length = 3)

check err here?

File: netsh/netsh.go [hunk 3]
```diff
@@ -124,6 +132,11 @@ func (runner *runner) GetInterfaces() ([]Ipv4Interface, error) {
 				currentInterface.DefaultGatewayAddress = value
 			}
 		}
+
+		// Assign Index based on map
+		if currentInterface != (Ipv4Interface{}) {
+			currentInterface.Idx = indexMap[currentInterface.Name]
```
Comment (rakelkar) [inline netsh/netsh.go]: >indexMap [](start = 26, length = 8)

check contains first.. and fail if not found

File: netsh/netsh.go [hunk 4]
```diff
@@ -138,6 +151,47 @@ func (runner *runner) GetInterfaces() ([]Ipv4Interface, error) {
...
+		return nil, err
+	}
+
+	// Split output by line
+	outputString := string(output[:])
+	outputString = strings.TrimSpace(outputString)
+	var outputLines []string = strings.Split(outputString, "\n")
+
+	if len(outputLines) < 3 {
+		return nil, errors.New("netsh bad")
```
Comment (rakelkar) [inline netsh/netsh.go]: >netsh bad [](start = 26, length = 9)

eh?

File: netsh/netsh.go [hunk 5]
```diff
@@ -138,6 +151,47 @@ func (runner *runner) GetInterfaces() ([]Ipv4Interface, error) {
...
+	}
+
+	// Remove first two lines of header text
+	outputLines = outputLines[2:]
+
+	indexMap := make(map[string] int)
+
+	reg := regexp.MustCompile("\\s{2,}")
```
Comment (rakelkar) [inline netsh/netsh.go]: >MustCompile [](start = 15, length = 11)

use previous regex to account for multiple spaces in a quoted string - and also test?

**Comment by rakelkar** [comment]
> 				if val, err := strconv.Atoi(match[1]); err == nil {

separate impl from join

---
Refers to: netsh/netsh.go:126 in 7f904d2. [](commit_id = 7f904d28d78d6b262c24cb15dfc7a2c5422e3dd8, deletion_comment = False)

File: netsh/netsh.go (hunk 5 above)
Comment (jroggeman) [reply netsh/netsh.go]: Will add tests; output of netsh int ipv4 show int doesn't quote the names though

File: netsh/netsh.go [hunk 6]
```diff
@@ -68,8 +70,37 @@ func New(exec utilexec.Interface) Interface {
...
+	}
+
+	// zip them up
+	for _, inter := range interfaces {
+		name := inter.Name
+
+		if val, ok := indexMap[name]; ok {
+			inter.Idx = val
+		} else {
+			fmt.Errorf("no index found for interface \"%v\"", name)
```
Comment (rakelkar) [inline netsh/netsh.go]: >fmt.Errorf("no index found for interface \"%v\"", name) [](start = 3, length = 55)

did you mean to return this error?

File: netsh/netsh.go [hunk 7]
```diff
@@ -85,6 +116,11 @@ func (runner *runner) GetInterfaces() ([]Ipv4Interface, error) {
 	var currentInterface Ipv4Interface
 	quotedPattern := regexp.MustCompile("\\\"(.*?)\\\"")
 	cidrPattern := regexp.MustCompile("\\/(.*?)\\ ")
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (rakelkar) [inline netsh/netsh.go]: what err would that be?

File: netsh/netsh.go [hunk 8]
```diff
@@ -68,8 +70,37 @@ func New(exec utilexec.Interface) Interface {
 	return runner
 }
 
-// GetInterfaces uses the show addresses command and returns a formatted structure
 func (runner *runner) GetInterfaces() ([]Ipv4Interface, error) {
+	// get interfaces
+	interfaces, err := runner.getIpAddressConfigurations()
+
+	if err != nil {
```
Comment (rakelkar) [inline netsh/netsh.go]: gofmt this file?

File: netsh/netsh.go [hunk 9]
```diff
@@ -138,6 +174,50 @@ func (runner *runner) GetInterfaces() ([]Ipv4Interface, error) {
 	return interfaces, nil
 }
 
+func getNetworkInterfaceParameters(runner* runner) (map[string]int, error) {
```
Comment (rakelkar) [inline netsh/netsh.go]: >(runner* runner [](start = 34, length = 15)

why not take as this param like everywhere else?

File: netsh/netsh.go (hunk 6 above)
Comment (jroggeman) [reply netsh/netsh.go]: hah, thought this would format and return the error in one step. yeah meant to return

File: netsh/netsh.go (hunk 9 above)
Comment (jroggeman) [reply netsh/netsh.go]: gogland yelled at me for having a private func as part of interface, so did this instead.  will try more experimentation to see if i can fix

File: netsh/netsh.go (hunk 7 above)
Comment (jroggeman) [reply netsh/netsh.go]: Assumed this was common pattern for propagating errors. Can rename and use specific values for each error case

File: netsh/netsh.go (hunk 9 above)
Comment (jroggeman) [reply netsh/netsh.go]: it didn't yell at me this time so who knows what I was doing before 🤷‍♂️

File: netsh/netsh.go (hunk 8 above)
Comment (jroggeman) [reply netsh/netsh.go]: Ran through gogland code reformat

File: netsh/netsh.go (hunk 7 above)
Comment (jroggeman) [reply netsh/netsh.go]: it didn't autoclose this but the errors are renamed in latest commit

**Pull request merged**